numpyro/infer/util.py

```diff
@@ -61,16 +61,20 @@ def log_density(model, model_args, model_kwargs, params, skip_dist_transforms=Fa
         if site['type'] == 'sample':
             value = site['value']
             intermediates = site['intermediates']
+            mask = site['mask']
+            scale = site['scale']
+            # exit early if mask all
+            if not np.any(mask):
```

Comment:
I think you would need to check `not_jax_tracer` here.

Reply:
Thanks for noting this; I think these optimizations need to be under `lax.cond` if we want to apply them under JIT. I'll also add a test for this.

Reply:
I think lax.cond does not have grad rule yet. I think you can just multiply for now.

Reply:
I see, okay let me run some investigations at my end and update.

Reply:
So I added some of these optimizations back by using `None` as the default value for mask and scale. Also added a test for jit. I have removed this particular early exit option, since we currently don't need it, i.e. we don't have use cases where we put `mask=False` (we can use `block` if needed). However, as I mentioned, this may not work correctly for dynamic masks under JIT, i.e. if for the same site, the mask changes from `None` to some bool array, this will give a wrong answer under JIT, but I think the only way to avoid that is to not apply any of these optimizations.